### PR TITLE
Sys_Error implementation

### PR DESCRIPTION
**Pull request by shit-ware**
Rewrote Sys_Error and some functions it depends on.

### CODE REVIEW

File: src/game/Cbuf/Cbuf.cpp [hunk 1]
```diff
@@ -53,4 +53,38 @@ int Conbuf_CleanText(const char* source, char* target, int sizeofTarget)
 
 	*target = 0;
 	return target - start;
+}
+
+//THUNK : 0x004F5770
```
Comment (mjkzy) [inline src/game/Cbuf/Cbuf.cpp:58]: thunk?

File: src/game/Cbuf/Cbuf.cpp [hunk 2]
```diff
@@ -53,4 +53,38 @@ int Conbuf_CleanText(const char* source, char* target, int sizeofTarget)
...
+long Conbuf_AppendText(const char* a1)
```
Comment (mjkzy) [inline src/game/Cbuf/Cbuf.cpp:59]: `long` -> `void`
`a1` can be labelled as `pMsg`

File: src/game/Cbuf/Cbuf.cpp [hunk 3]
```diff
@@ -53,4 +53,38 @@ int Conbuf_CleanText(const char* source, char* target, int sizeofTarget)
...
+{
+	char* v1;
```
Comment (mjkzy) [inline src/game/Cbuf/Cbuf.cpp:61]: v1 should be `const char*`

File: src/game/Cbuf/Cbuf.cpp [hunk 4]
```diff
@@ -53,4 +53,38 @@ int Conbuf_CleanText(const char* source, char* target, int sizeofTarget)
...
+	if (strlen(a1) <= 16383)
+	{
+		v1 = (char*)a1;
+	}
+	else
+	{
+		v1 = (char*)&a1[strlen(a1) - 16383];
+	}
+
+	v2 = (std::int32_t)memory::call<unsigned char*(char*, int, char*)>(0x64DD30)(lParam, 0x8000, v1);
```
Comment (mjkzy) [inline src/game/Cbuf/Cbuf.cpp:74]: the memory call goes to `Conbuf_CleanText`, which we wrote

File: src/game/Cbuf/Cbuf.cpp [hunk 5]
```diff
@@ -53,4 +53,38 @@ int Conbuf_CleanText(const char* source, char* target, int sizeofTarget)
...
+	*(std::int32_t*)0x64A38B8 += v2;
```
Comment (mjkzy) [inline src/game/Cbuf/Cbuf.cpp:75]: 0x64A38B8 is s_totalChars, and the type may be `unsigned int`

File: src/game/Cbuf/Cbuf.cpp [hunk 6]
```diff
@@ -53,4 +53,38 @@ int Conbuf_CleanText(const char* source, char* target, int sizeofTarget)
...
+	}
+	else
+	{
+		SendMessageA(*(HWND*)0x64A328C, 0xB1, 0, -1);
+		*(std::int32_t*)0x64A38B8 = v2;
+	}
+
+	SendMessageA(*(HWND*)0x64A328C, 0xB6, 0, 0xFFFF);
+	SendMessageA(*(HWND*)0x64A328C, 0xB7, 0, 0);
+	return SendMessageA(*(HWND*)0x64A328C, 0xC2, 0, (long)lParam);
```
Comment (mjkzy) [inline src/game/Cbuf/Cbuf.cpp:89]: no return since its a void

File: src/game/Render/Render.cpp [hunk 7]
```diff
@@ -77,3 +77,31 @@ bool R_PushRemoteScreenUpdate(std::int32_t a1)
     }
     return IsMainThread;
 }
+
+//DONE : 0x004CFA00
+std::int32_t FixWindowsDesktop()
```
Comment (mjkzy) [inline src/game/Render/Render.cpp:82]: `std::int32_t` -> `void`

File: src/game/Render/Render.cpp [hunk 8]
```diff
@@ -77,3 +77,31 @@ bool R_PushRemoteScreenUpdate(std::int32_t a1)
...
+{
+    unsigned long currentThreadId = GetCurrentThreadId();
```
Comment (mjkzy) [inline src/game/Render/Render.cpp:84]: this should be `DWORD`, not `unsigned long`

File: src/game/Render/Render.cpp [hunk 9]
```diff
@@ -77,3 +77,31 @@ bool R_PushRemoteScreenUpdate(std::int32_t a1)
...
+    HWND desktopWindow = GetDesktopWindow();
+    HDC DC = GetDC(desktopWindow);
+    std::int32_t v3 = 0;
+    std::int8_t* v4;
+    std::int32_t v5 = 256;
+    void* ramp; //char[512]..? never used according to ida
```
Comment (mjkzy) [inline src/game/Render/Render.cpp:90]: this is oddly defined, no clue if this is psuedocode going nuts
`unsigned __int16 ramp[3][256];`

and `v4` is defined as `unsigned __int16*`, and its `v4 = ramp[1]`? idk its a bunch of nonsense

File: src/game/Render/Render.cpp [hunk 10]
```diff
@@ -77,3 +77,31 @@ bool R_PushRemoteScreenUpdate(std::int32_t a1)
...
+        *((std::uint16_t*)v4 - 256) = v3;
+        *(std::uint16_t*)v4 = v3;
+        *((std::uint16_t*)v4 + 256) = v3;
+        v3 += 257;
+        v4 += 2;
+        v5--;
+    } 
+    while (v5);
+    SetDeviceGammaRamp(DC, ramp);
+    return ReleaseDC(desktopWindow, DC);
```
Comment (mjkzy) [inline src/game/Render/Render.cpp:106]: no return since void

File: src/game/Sys/Sys.cpp [hunk 11]
```diff
@@ -401,4 +392,75 @@ bool Sys_DatabaseCompleted()
...
+    memory::call<std::int32_t(void)>(0x48A4E0)(); //Steam_EmergencyShutdown?
+    exit(0);
+
+}
+
+//DONE : 0x00433940
+std::int32_t Sys_SetErrorText(const char* text)
+{
+    HWND activeWindow = GetActiveWindow();
+    I_strncpyz(*(char**)0x64A329C, (char*)text, 512);
```
Comment (mjkzy) [inline src/game/Sys/Sys.cpp:436]: pretty sure this doesn't need a `(char*)` cast

File: src/game/Sys/Sys.cpp [hunk 12]
```diff
@@ -401,4 +392,75 @@ bool Sys_DatabaseCompleted()
...
+        while (GetMessageA(&msg, 0, 0, 0));
+    }
+    va_end(args);
+    memory::call<std::int32_t(void)>(0x48A4E0)(); //Steam_EmergencyShutdown?
+    exit(0);
+
+}
+
+//DONE : 0x00433940
+std::int32_t Sys_SetErrorText(const char* text)
```
Comment (mjkzy) [inline src/game/Sys/Sys.cpp:433]: `std::int32_t` -> `void`

File: src/game/Sys/Sys.cpp [hunk 13]
```diff
@@ -401,4 +392,75 @@ bool Sys_DatabaseCompleted()
...
+{
+    HWND activeWindow = GetActiveWindow();
+    I_strncpyz(*(char**)0x64A329C, (char*)text, 512);
+    DestroyWindow(*(HWND*)0x64A3298);
+    *(HWND*)0x64A3298 = 0;
+    return MessageBoxA(activeWindow, text, "Error", 0x10);
```
Comment (mjkzy) [inline src/game/Sys/Sys.cpp:439]: no return since void

**Review by mjkzy — CHANGES_REQUESTED** [review]

File: src/game/Cbuf/Cbuf.cpp (hunk 1 above)
Comment (shit-ware) [reply src/game/Cbuf/Cbuf.cpp:58]: Still waiting on clarification on when to use those.

File: src/game/Cbuf/Cbuf.cpp (hunk 4 above)
Comment (shit-ware) [reply src/game/Cbuf/Cbuf.cpp:74]: My IDA instance doesn't always apply symbols from Lumina and I don't know why, otherwise I would have labeled it

Comment (mjkzy) [reply src/game/Cbuf/Cbuf.cpp:74]: me neither, so I just compare it to T6 Dedi IDB. Sys_* code doesn't seem to change that much between CODs so a good guess can be made

File: src/game/Render/Render.cpp (hunk 8 above)
Comment (shit-ware) [reply src/game/Render/Render.cpp:84]: DWORD is a macro, they mean the same thing, we are trying (keyword is "trying") to maintain compatibility across tool sets.
image

File: src/game/Render/Render.cpp (hunk 9 above)
Comment (shit-ware) [reply src/game/Render/Render.cpp:90]: Yeah, I have no idea either, certified IDA moment.

**Pull request merged**